shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.mm

```diff
@@ -0,0 +1,119 @@
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.h"
+
+#include <OpenGL/gl.h>
+
+enum {
+  kFront = 0,
+  kBack = 1,
+  kBufferCount,
+};
+
+@interface FlutterSurfaceManager () {
+  CGSize surfaceSize;
+  CALayer* layer;
+  NSOpenGLContext* openGLContext;
+  uint32_t _frameBufferId[kBufferCount];
+  uint32_t _backingTexture[kBufferCount];
+  IOSurfaceRef _ioSurface[kBufferCount];
+}
+@end
+
+@implementation FlutterSurfaceManager
+
+- (instancetype)initWithLayer:(CALayer*)layer_ openGLContext:(NSOpenGLContext*)opengLContext_ {
+  if (self = [super init]) {
+    layer = layer_;
+    openGLContext = opengLContext_;
+
+    NSOpenGLContext* prev = [NSOpenGLContext currentContext];
```

Comment:
We are doing this pattern of

1. Save old context.
2. Make context current.
3. Reset old context or clear it.

This is being done at a couple of places in this class. This is very reminiscent of `GLContextSwitch`. Can we use that wrapper here or create an RAII wrapper?

Reply:
Added [MacOSSwitchableOpenGLContext](https://github.com/flutter/engine/pull/21525/commits/4873b83521782c7c6e8ca2cab113269ed2b6850e).